src/connect/planner.cpp

```diff
@@ -690,17 +715,60 @@ void Planner::background_done(BackgroundResult result) {
     background_command.reset();
 }
 
-void Planner::download_done() {
+void Planner::download_done(DownloadStep result) {
     // Similar reasons as with background_done
     assert(download.has_value());
-    // We do _not_ set the event here. We do so in watching the transfer.
-    //
-    // But we make sure the observed_transfer is set even if there was no
-    // next_event in the meantime or if it was short-circuited.
+    if (result == DownloadStep::FailedNetwork && download->allowed_retries > 0) {
+        assert(!download->need_retry);
+        download->allowed_retries--;
+        download->need_retry = true;
+    } else {
+        // We do _not_ set the event here. We do so in watching the transfer.
+        //
+        // But we make sure the observed_transfer is set even if there was no
+        // next_event in the meantime or if it was short-circuited.
+
+        observed_transfer = Monitor::instance.id();
+        assert(observed_transfer.has_value()); // Because download still holds the slot.
+        download.reset();
+    }
+}
+
+void Planner::recover_download() {
+    assert(download.has_value());
+    assert(download->need_retry);
+    download->need_retry = false;
+    uint32_t position = download->download.position();
+
+    const auto [config, config_changed] = printer.config(false);
+    const auto [host, port] = host_and_port(config, download->port);
 
-    observed_transfer = Monitor::instance.id();
-    assert(observed_transfer.has_value()); // Because download still holds the slot.
-    download.reset();
+    char url[enc_url_len];
+    make_enc_url(url, download->orig_iv);
+
+    char range[6 /* bytes = */ + 10 /* 2^32 in text */ + 1 /* - */ + 1 /* \0 */];
+    snprintf(range, sizeof range, "bytes=%" PRIu32 "-", position);
+    HeaderOut hdrs[2] = {
+        { "Range", range, nullopt },
+        { nullptr, nullptr, nullopt },
```

Comment:
Just a thought, not really related to this PR, but maybe it would be nice to have this "null" terminating line added automatically and not having to put it in every extra header out there :thinking: .

Reply:
Any suggestion on _how_? I mean, I could just _omit_ it and rely on the fact there's an extra element in the array and that it would be auto-zeroed, but that IMO hurts readability and resilience of the code against changes (if you add another header and this doesn't fit, it should IMO fail to compile instead of doing some random walking of stack memory).

Reply:
My thought was to automatically add the line, while actually sending the request. So adding it somehow in the `send_request` function and the caller wouldn't need to worry about it at all, since it is needed every single time.

Reply:
That line is there to mark the end of the array. We either need a „sentinel entry“ or a size. At that time, sentinel value looked nicer.

That is, the function doesn't _need_ an empty header for an actual purpose, it needs to know where the headers end. Adding it automatically doesn't work, since the function doesn't know where ‒ and if it did, it wouldn't have to add it, as there would be no further purpose.